### ISSUE TITLE
disconnect method

### ISSUE THREAD

**Issue by jimbasquiat** [erdewit/ib_insync#10]
This is probably a silly question so apologize in advance..:
I cannot seem to be able to have the disconnect method work - also, when looking into the source code I am not sure how it should work:
In the ib.py source file, the .disconnect() method is an overload from the .disconnect() in client.py.
In client.py, looking for the method using ctrl-f i find it in the Connection class. The Connection class is in turn referenced in the client proper in the connect_async method, but is declared as self.conn = Connection(host, port). This should be that the deconnect() method is at self,conn.disconnect()?
For me using 
tws=Client()
tws.connect(.....)
tws.disconnect() # doesnt work
tws.conn.disconnect() # also doesnt work

I'm using the jupyter notebook, so it is somewhat more important to be able to properly explicitely disconnect()

**Comment by erdewit** [comment]
I see that you are using the Client directly, instead of the IB component. This should be no problem. If in a notebook I try this:

```python
from ib_insync import Client
from ibapi.wrapper import EWrapper

class TestClient(Client, EWrapper):
    def __init__(self):
        Client.__init__(self, wrapper=self)

    def managedAccounts(self, accountsList):
        print(self.__class__.__name__, accountsList)

client = TestClient()
client.connect(host='127.0.0.1', port=7497, clientId=1)
print(client.isConnected())
client.disconnect()
print(client.isConnected())
```

then the disconnect() works fine. What sort of code are you using and what errors do you get?

**Comment by jimbasquiat** [comment]
Hi Ewald, sorry for my delayed response.

This is especially a problem or me in jupyter. 
I can connect with:
```
tws=TWSClient() #(the name of the client in my code)
tws.connect()
```

This works fine and i can then run the various reuests I need.

However when I type
`tws.diconnect()`

I can see by glancing at the ib gateway console that the client is still connected. No error meassage is returned though.

I am using a "simplified" version of your project (actually an old version from April i think that i customised for my needs) I like that there is less code to navigate.

In my version the code for the connection is:
```
    @async_loop
    async def connect(self, host="127.0.0.1", port=4002, clientId=1):
        self.host = host
        self.port = port
        self.clientId = clientId
        self.idfuture=asyncio.Future()
        self.setConnState(EClient.CONNECTING)
        self.conn = TWSConnection(host, port)
        self.conn.connected = self._onSocketConnected
        self.conn.hasData = self._onSocketHasData
        self.conn.connect()
        await self.idfuture

The call back for connection is:
```
    def nextValidId(self, orderId):
        self.orderId = orderId
        self.idfuture.set_result('done')
```

async loop is simply a decorator that run the loop over the function:
```
    def async_loop(f):
        def decorated(self, *args, **kwargs):
                self.loop.run_until_complete(f(self, *args, **kwargs))
        return decorated
```

Now to the `disconnect()` method:

I didn't touch the code that use to be in that version of your project. It doesn't seems to be very different from the current one though. However in that version (and also in the current version of your code) I can only see a `disconnect()` method in the Connection class (in client.py). There is no `disconnect()` method in you Client class. The IB class (in ib.py) calls a `self.client.disconnect() `-> that would be the `disconnect()` from the original ib.api then no?

So here is the code that I have in my Connection class:
```
    def disconnect(self):
        if self.socket:
            self.socket.transport.close()
            self.socket = None 
```
Given that the `connect() `method calls the Connection class with:

```
        self.conn = TWSConnection(host, port)
        self.conn.connected = self._onSocketConnected
        self.conn.hasData = self._onSocketHasData
        self.conn.connect()
```
I think `disconnect()` should work with `self.conn.disconnect()`. But this doesnt work either.

Sorry my post is not synthetic but i think i am missing multiple things here as the absence of link between the asyncio disconnect() ( in connection class) and the client seems already wrong - that is why i am confused actually.

Anyway i'd hbe happy to exchange more with you as i find that issue difficult to explain.

**Comment by erdewit** [comment]
The ibapi EClient.disconnect method is used for disconnecting. You can trace this to see if, eventually, transport.close() is called on the socket (just add some print statements along the route for example). Since you're using old code from the tws_async project with modifications I can't really tell more about it.

**Comment by jimbasquiat** [comment]
[I found the solution here eventually](https://stackoverflow.com/questions/35872750/how-to-close-python-asyncio-transport).

Basically the `client.disconnect` method had to be re-written as 
```
async def disconnect(self):
    self.conn.disconnect()
    await asyncio.sleep(0)
```
as explained in the SO post.
I am surprised that your version of the ib-insync still can disconnect. Anyway, Hopefully this can be useful to somenone.

**Comment by erdewit** [comment]
Calling ``loop.run_until_complete(asyncio.sleep(0))`` will run one iteration of the event loop. If this makes things work that means that there was no event loop running to begin with.

**Comment by JoeFernando** [comment]
I'm brand new to python and have started using this library to download data from IBKR. I live in Sydney, Australia and faced the same time issue. 

The way I solved it was change the date input format:

ASX kick-off is 10am local time. 

when I used `start = datetime.datetime(2022, 10, 18, tzinfo=datetime.timezone.utc)` the download starts at 11am local time. 

Instead, I used `start = datetime.datetime(2022, 10, 17, 23, 00, 00, tzinfo=datetime.timezone.utc)` which gives the correct output from 10am local time.

Thanks for this library. This is way more easier to use than IBKR's!
